code-walktrhough.md

```diff
@@ -0,0 +1,206 @@
+
+## Guest code overview
+
+The guest code runs off-chain within the RISC Zero zkVM. It performs the actual computation to check if a given number is even, generating a proof of this computation. This code forms the basis for the off-chain computation part of the Ethereum zk-coProcessor.
+
+### Guest code walkthrough:
+
+```rust
+use std::io::Read;
+use alloy_primitives::U256;
+use alloy_sol_types::SolValue;
+use risc0_zkvm::guest::env;
+
+fn main() {
+    let mut input_bytes = Vec::<u8>::new();
+    env::stdin().read_to_end(&mut input_bytes).unwrap();
+    let number = <U256>::abi_decode(&input_bytes, true).unwrap();
+    assert!(number.bit(0) == false, "number is not even");
+    env::commit_slice(number.abi_encode().as_slice());
+}
+```
+
+- **Input reading**: The code starts by reading the input data, which in this case is the number whose evenness is to be verified. This number is passed into the zkVM as a byte array.
+- **Decoding the input**: The input bytes are decoded into an U256 type, which stands for an unsigned 256-bit integer. This is a common data type for representing numbers in Ethereum smart contracts due to its ability to handle large integers used in cryptographic operations.
+- **Verification logic**: The core of the guest code is a simple assertion that checks if the least significant bit (bit(0)) of the number is 0 (indicating the number is even). If the number is not even, the assertion fails, and the zkVM will not produce a valid proof.
+- **Committing the Journal**: If the assertion passes, the code commits to the journal (i.e., the output of the computation). The journal contains the encoded version of the number, prepared in such a way that the smart contract can easily decode it using Solidity's ABI (Application Binary Interface) decoding functions.
+
+## Smart Contract overview
+
+While the RISC Zero zkVM enables developers to write off-chain logic as guest code, zkVM proofs (specifically a Groth16 SNARKs) get verified and incorporated into smart contract logic on chain. The end result is that the guest logic which checks if a number is even, is performed off chain but enforced on chain. This model extends to any program, however complex, which can be written as a guest program of the RISC Zero zkVM via Rust.
+The EvenNumber smart contract verifies a RISC Zero zkVM proof that a number is even.
+In this simple example, if a proof of an even number is verified, we will set that even number in the smart contract's state.
+
+### Key Components:
+- **RISC Zero Verifier**: An interface to the RISC Zero verifier contract, which checks the validity of a zero-knowledge proof. The actual verifier contract would be deployed independently, and its address is passed to the `EvenNumber` contract upon initialization.
```

Comment:
I don't really think the second sentence is helping to increase clarity. This is somewhat of an implementation detail, which is handled in the deploy script and is clear from reading the code below.
```suggestion
- **RISC Zero Verifier**: An interface to the RISC Zero verifier contract, which checks the validity of a zero-knowledge proof.
```